src/main/java/org/wildfly/security/auth/server/MechanismConfiguration.java

```diff
@@ -41,14 +43,19 @@
  * @author <a href="mailto:david.lloyd@redhat.com">David M. Lloyd</a>
  */
 public final class MechanismConfiguration {
+
+    private final MechanismConfiguration parent;
```

Comment:
The alternative is I introduce a new outer class as a wrapper that holds a reference to the default MechanismConfiguration as well as the Map<serverName, MechanismConfiguration> for the server specific configurations, this will need passing to the ServerAuthenticationContext which would then have to hold a reference to this new class as well as pulling out the default in case no server specific is specified.

If we stick with the Maps of Maps then I think the key order of mech name followed by host name makes more sense - the main motivation for these changes was to get to a single SecurityFactory<Credential> and as typing of this Credential is very tied to the mechanism keying on the mechanism first made more sense - keying on host name was then for specific instances for that host.

So if we do want to switch to some wrapper around the MechanismConfiguration instances anyway it could make sense that we do have a MechanismConfigurationSelector - regarding the values passed to the selector I would think start with mechanism name, host name, and protocol - our InitialState at the very least would have to hold onto these three values and convert to a MechanismConfiguration before transitioning.  Another advantage of this as well is each mechanism does not have to be defined in the configuration to be usable - if the MechanismConfigurationSelector can return a MechanismConfiguration for that combination then the mechanism is usable - but then the risk is it is very much later in the authentication process that we discover the mechanism can not be used.

But the issue here is I am trying to get HTTP authentication complete with a full set of mechanisms, diverting to look at this is because we have a current hole that does not quite satisfy SPNEGO / Kerberos authentication - this is fast becoming an even bigger redesign of MechanismConfiguration selection other than allowing host specific config with single SecurityFactory references.

Will think on this again in the morning.